Implement offsetParent/Top/Left/Width/Height.

glennw opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6784)

<!-- Reviewable:end -->

gw3583 committed:
Implement offsetParent/Top/Left/Width/Height.

glennw commented:
This PR fixes a large number of layout issues in the enyojs sampler. But I'm not sure if the implementation could be made a lot simpler, it feels a bit complicated. Comments?

r? @jdm for script bits and @pcwalton for the layout bits?

pcwalton commented on components/layout/layout_task.rs:
```diff
@@ -127,6 +127,9 @@ pub struct LayoutTaskData {
 
+    /// A queued response for the offset parent/rect of a node.
+    pub offset_parent_response: OffsetParentResponse,
+
```
This is probably going to get unwieldy over time. Perhaps we should factor this out into a struct now?

glennw replied:
@jdm opened https://github.com/servo/servo/issues/6787 - perhaps we can do this as a follow up?

jdm replied:
It is, in fact, prime E-easy material :)

pcwalton commented on components/layout/flow.rs:
```diff
@@ -284,6 +284,7 @@ pub trait Flow: fmt::Debug + Sync {
     /// Iterates through border boxes of all of this flow's fragments.
     fn iterate_through_fragment_border_boxes(&self,
                                              iterator: &mut FragmentBorderBoxIterator,
+                                             level: i32,
```
Add a comment documenting the level parameter.

pcwalton commented on components/layout/layout_task.rs:
```diff
@@ -1362,7 +1399,7 @@ impl UnioningFragmentBorderBoxIterator {
 
 impl FragmentBorderBoxIterator for UnioningFragmentBorderBoxIterator {
-    fn process(&mut self, _: &Fragment, border_box: &Rect<Au>) {
+    fn process(&mut self, _: &Fragment, _: i32,border_box: &Rect<Au>) {
```
uber-nit: space after `i32,` for consistency

pcwalton commented on components/script/dom/window.rs:
```diff
@@ -769,6 +770,28 @@ impl<'a> WindowHelpers for &'a Window {
+        self.reflow(ReflowGoal::ForScriptQuery,
+                    ReflowQueryType::OffsetParentQuery(node),
+                    ReflowReason::Query);
+        self.join_layout(); //FIXME: is this necessary, or is layout_rpc's mutex good enough?
```
I think it's not necessary

glennw commented:
Ref https://github.com/servo/servo/issues/6643

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6771) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
@pcwalton Addressed those comments and rebased.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit b91c0a0 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton b91c0a04b4f365b4cba427d99dd2270dcf36f83a -->

bors-servo commented:
:hourglass: Testing commit b91c0a0 with merge f6cdf7b...

bors-servo commented:
:broken_heart: Test failed - [mac3](http://build.servo.org/builders/mac3/builds/682)

metajack commented:
```
/css21_dev/html4/first-letter-dynamic-003a.htm
----------------------------------------------
FAIL [Parent]
/css21_dev/html4/first-letter-dynamic-003b.htm
----------------------------------------------
FAIL [Parent]
/css21_dev/html4/run-in-basic-017.htm
-------------------------------------
FAIL [Parent]
/css21_dev/html4/run-in-relpos-between-003.htm
----------------------------------------------
CRASH expected FAIL [Parent]
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6893) made this pull request unmergeable. Please resolve the merge conflicts.

gw3583 committed:
Fix panic when called on root node, and update test expectations.

The three tests that are marked as failures are because they use offsetWidth as a dummy to force a layout flush. Now that these exist, they expose an unrelated bug with reflow.

glennw commented:
@pcwalton re-r? Fixed a panic in the css tests and updated expectations. If you're fine with that change, I'll squash and rebase after that.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit ade6765 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton ade676527d64bdcb7449b0a9170e21cc046c6e98 -->

bors-servo commented:
:hourglass: Testing commit ade6765 with merge 230a8a4...

bors-servo commented:
:broken_heart: Test failed - [mac1](http://build.servo.org/builders/mac1/builds/2806)

glennw commented:
@bors-servo retry

jdm commented:
@bors-servo: r-
http://mxr.mozilla.org/servo/source/tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/input-type-button.html?force=1#35
That test is legitimately passing now. Please look at failures that occur?

glennw commented:
@jdm Yep, I just managed to repro it locally, so will look into it.

glennw commented:
@jdm Ah, I actually misread that build log - there was a subtest in input-type-button that was previously failing, and this PR fixes that subtest. I have updated the last commit to remove that test failure from the WPT metadata.

jdm commented:
That ini file can be removed now, if it contains no other failures.

glennw commented:
Updated to remove that ini file.

glennw commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit 39972a1 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 39972a174e5b40966307ca32fa38325cc7fddbd0 -->

bors-servo commented:
:hourglass: Testing commit 39972a1 with merge d66c59a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2557), [gonk](http://build.servo.org/builders/gonk/builds/1706), [linux1](http://build.servo.org/builders/linux1/builds/2558), [linux2](http://build.servo.org/builders/linux2/builds/2576), [linux3](http://build.servo.org/builders/linux3/builds/716), [mac1](http://build.servo.org/builders/mac1/builds/2810), [mac2](http://build.servo.org/builders/mac2/builds/2792), [mac3](http://build.servo.org/builders/mac3/builds/750)
